vj4/ui/components/ide/ide.js

```diff
@@ -0,0 +1,448 @@
+      i += 1;
+    }
+    return ret;
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      displayTesting: false,
+      displayEvalutating: false,
+      code: '',
+      currentTestingInputText: '',
+      currentTestingOutputText: '',
+      language: 'c',
+      testingData: [],
+      resultData: [],
+      currentActiveTabIndex: -1,
+    };
+    this.sock = null;
+  }
+
+  componentDidMount() {
+    if (!this.sock) {
+      this.sock = sockJS('/records-conn');
+      const self = this;
+      this.sock.onmessage = (message) => {
+        const nextResultData = self.state.resultData.slice();
+        const rdoc = JSON.parse(message.data).rdoc;
+        const existedRdoc = nextResultData.filter(val => val._id === rdoc._id)[0];
+        if (existedRdoc) {
+          const idx = nextResultData.indexOf(existedRdoc);
+          nextResultData[idx] = rdoc;
+        } else {
+          nextResultData.unshift(rdoc);
+        }
+        return self.setState({
+          resultData: nextResultData,
+        });
+      };
+    }
+  }
+
+  onLeaveIDEMode() {
+    if (window.confirm('确认退出IDE模式吗？')) {
+      $('body').removeClass('ide-mode');
+      window.location.hash = '';
+    }
+  }
+
+  onLanguageChange(ev) {
+    // console.log(this.refs.editor.getCodeMirror());
+    this.setState({
+      language: ev.target.value,
+    });
+    this.refs.editor.getCodeMirror().setOption('mode', Ide.mapLangToMode[ev.target.value]);
+  }
+
+  onCodeUpdate(newCode) {
+    this.setState({
+      code: newCode,
+    });
+  }
+
+  onAddTestCase(ev, cb = () => {}) {
+    const newTestcase = {
+      input: '',
+      output: '',
+    };
+    const nextTestingData = _.cloneDeep(this.state.testingData);
+    nextTestingData.push(newTestcase);
+    this.setState({
+      currentTestingInputText: '',
+      currentTestingOutputText: '',
+      testingData: nextTestingData,
+      currentActiveTabIndex: nextTestingData.length - 1,
+    }, cb);
+  }
+
+  onDeleteTestCase() {
+    const nextTestingData = _.cloneDeep(this.state.testingData);
+    nextTestingData.splice(this.state.currentActiveTabIndex, 1);
+    let nextActiveTabIndex = 0;
+
+    if ((this.state.currentActiveTabIndex === 0) && (this.state.testingData.length !== 1)) {
+      nextActiveTabIndex = 0;
+    } else if ((this.state.currentActiveTabIndex === 0) && (this.state.testingData.length === 1)) {
+      return this.setState({
+        currentTestingInputText: '',
+        currentTestingOutputText: '',
+        testingData: [],
+        currentActiveTabIndex: -1,
+      });
+    } else {
+      nextActiveTabIndex = this.state.currentActiveTabIndex - 1;
+    }
+    return this.setState({
+      testingData: nextTestingData,
+      currentActiveTabIndex: nextActiveTabIndex,
+      currentTestingInputText: nextTestingData[nextActiveTabIndex].input,
+      currentTestingOutputText: nextTestingData[nextActiveTabIndex].output,
+    });
+  }
+
+  onClickTestCaseTab(idx) {
+    return () => {
```

Comment:
The name might be confusing, my mistake.
See line 246: `onClickTestCaseTab(i)` returns a callback function for mouse clicking event.